Hw15/Vector2d.cpp

```diff
@@ -0,0 +1,91 @@
+#include "Vector2d.h"
+#include <cmath>
+
+Vector2d::Vector2d(const float x, const float y):
+m_x{x},
+m_y{y}
+{
+}
+
+Vector2d& Vector2d::operator=(const Vector2d& v)
+{
+	if (this == &v)
+		return *this;
+
+	m_x = v.m_x;
+	m_y = v.m_y;
+
+	return *this;
+}
+
+Vector2d Vector2d::operator+(const Vector2d& secondVector) const
+{
+	Vector2d v(*this);
+
+	v.m_x += secondVector.m_x;
+	v.m_y += secondVector.m_y;
+
+	return v;
+}
+
+Vector2d Vector2d::operator-(const Vector2d& secondVector) const
+{
+	Vector2d v(*this);
+
+	v.m_x -= secondVector.m_x;
+	v.m_y -= secondVector.m_y;
+
+	return v;
+}
+
+Vector2d operator+(const Vector2d& leftVector, const Vector2d& rightVector)
+{
+	Vector2d leftVector1(leftVector);
+	return leftVector1 += rightVector;
+}
+
+Vector2d operator-(const Vector2d& leftVector, const Vector2d& rightVector)
+{
+	Vector2d leftVector1(leftVector);
+	return leftVector1 -= rightVector;
+}
+
+Vector2d& Vector2d::operator+=(const Vector2d& rightVector)
+{
+	m_x += rightVector.m_x;
+	m_y += rightVector.m_y;
+
+	return *this;
+}
+
+Vector2d& Vector2d::operator-=(const Vector2d& rightVector)
+{
+	m_x -= rightVector.m_x;
+	m_y -= rightVector.m_y;
+
+	return *this;
+}
+
+float Vector2d::operator()() const
+{
+	const float res = static_cast<float>(std::sqrt(std::pow(m_x, 2) + std::pow(m_y, 2)));
+	return res;
+}
+
+float& Vector2d::operator[](const size_t i)
+{
+	return (&m_x)[i];
```

Comment:
that's very insecure way.
It will backfire if i is not in range 0-1
it may also break if types of m_x, m_y change due to memory alignment optimization possibility by complier

Reply:
"that's very insecure way." - Yep
"It will backfire if i is not in range 0-1" - Yep, but operator[] usually does not make any checks, like it is in vector or map. Just rapid access.
"it may also break if types of m_x, m_y change due to memory alignment optimization possibility by compiler" - Agree, that pointers arithmetic should be avoided.

Reply:
And not just this pointers arithmetic, but any pointers arithmetic should be avoided.